### PR TITLE
feat(framework) Add end-to-end FAB delivery

### PR DESCRIPTION
**Pull request by charlesbvll**
### Checklist

- [ ] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.ai/docs/writing-documentation.html)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`)

### CODE REVIEW

File: src/py/flwr/cli/run/run.py [hunk 1]
```diff
@@ -168,6 +168,8 @@ def on_channel_state_change(channel_connectivity: str) -> None:
     content = fab_path.read_bytes()
     fab = Fab(hashlib.sha256(content).hexdigest(), content)
 
+    fab = Fab("", Path(fab_path).read_bytes())
+
```
Comment (charlesbvll) [inline src/py/flwr/cli/run/run.py:172]:
```suggestion
```

File: src/py/flwr/client/app.py [hunk 2]
```diff
@@ -406,9 +403,17 @@ def _on_backoff(retry_state: RetryState) -> None:
...
+                        fab = get_fab(run.fab_hash)
+                    else:
+                        fab = None
+
                     # Register context for this run
                     node_state.register_context(
-                        run_id=run_id, run=runs[run_id], flwr_path=flwr_path
+                        run_id=run_id,
+                        default_config=get_fab_config(fab.content) if fab else {},
+                        run=runs[run_id],
```
Comment (danieljanes) [inline src/py/flwr/client/app.py:416]:
```suggestion
                        run=run,
```

File: src/py/flwr/server/run_serverapp.py [hunk 3]
```diff
@@ -162,10 +166,23 @@ def run_server_app() -> None:  # pylint: disable=too-many-branches
...
-        config = get_project_config(app_path)
+        if run_.fab_hash:
+            fab_req = GetFabRequest(hash_str=run_.fab_hash)
+            # pylint: disable-next=W0212
+            fab_res: GetFabResponse = driver._stub.GetFab(fab_req)
+            if fab_res.fab.hash_str != run_.fab_hash:
+                raise ValueError("FAB hashes don't match.")
+
+            config = get_fab_config(fab_res.fab.content)
+            server_app_run_config = fuse_dicts(config, run_.override_config)
```
Comment (danieljanes) [inline src/py/flwr/server/run_serverapp.py:178]: This looks like it downloads the FAB only for the default config, right? Shouldn't it also install the FAB and then later load the ServerApp from it?

File: src/py/flwr/server/superlink/fleet/vce/vce_api.py [hunk 4]
```diff
@@ -79,7 +79,7 @@ def _register_node_states(
 
         # Pre-register Context objects
         node_states[node_id].register_context(
-            run_id=run.run_id, run=run, app_dir=app_dir
+            run_id=run.run_id, default_config={}, run=run, app_dir=app_dir
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/fleet/vce/vce_api.py:82]: In `flwr/client/app.py`, we're not passing the `flwr_dir` any more. Should we remove `app_dir` here as well?

**Review by danieljanes — APPROVED** [review]

**Pull request merged**